cli/command/stack/kubernetes/services.go

```diff
@@ -2,43 +2,136 @@ package kubernetes
 
 import (
 	"fmt"
+	"sort"
+	"strings"
 
 	"github.com/docker/cli/cli/command/formatter"
 	"github.com/docker/cli/cli/command/stack/options"
 	"github.com/docker/cli/kubernetes/labels"
+	"github.com/docker/docker/api/types/filters"
+	appsv1beta2 "k8s.io/api/apps/v1beta2"
+	corev1 "k8s.io/api/core/v1"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+var supportedServicesFilters = map[string]bool{
+	"mode":  true,
+	"name":  true,
+	"label": true,
+}
+
+func generateValuedLabelsSelector(valuedLabels map[string][]string) []string {
+	var result []string
+	for k, v := range valuedLabels {
+		if len(v) == 1 {
+			result = append(result, fmt.Sprintf("%s=%s", k, v[0]))
+		} else {
+			sort.Strings(v)
+			result = append(result, fmt.Sprintf("%s in (%s)", k, strings.Join(v, ",")))
```

Comment:
is the `in (x)` operator less efficient than `x=y`, or will the selector be smart enough to optimise if there's a single value?

If so, we could simplify to;

```go
for k, v := range valuedLabels {
    sort.Strings(v)
    result = append(result, fmt.Sprintf("%s in (%s)", k, strings.Join(v, ",")))
}
```

Reply:
I honestly don't know if it's really less efficient (we are talking about 1 equality vs a list of 1...)